core/service/src/lib.rs

```diff
@@ -914,38 +923,40 @@ where
 		transactions_to_propagate(&self.pool)
 	}
 
-	fn import(&self, transaction: &<B as BlockT>::Extrinsic) -> Option<H> {
+	fn hash_of(&self, transaction: &B::Extrinsic) -> H {
+		self.pool.hash_of(transaction)
+	}
+
+	fn import(&self, report_handle: ReportHandle, who: PeerId, reputation_change: i32, transaction: B::Extrinsic) {
 		if !self.imports_external_transactions {
 			debug!("Transaction rejected");
-			return None;
+			return;
 		}
 
 		let encoded = transaction.encode();
 		match Decode::decode(&mut &encoded[..]) {
 			Ok(uxt) => {
 				let best_block_id = BlockId::hash(self.client.info().chain.best_hash);
-				match self.pool.submit_one(&best_block_id, uxt) {
-					Ok(hash) => Some(hash),
-					Err(e) => match e.into_pool_error() {
-						Ok(txpool::error::Error::AlreadyImported(hash)) => {
-							hash.downcast::<H>().ok()
-								.map(|x| x.as_ref().clone())
-						},
-						Ok(e) => {
-							debug!("Error adding transaction to the pool: {:?}", e);
-							None
-						},
-						Err(e) => {
-							debug!("Error converting pool error: {:?}", e);
-							None
-						},
-					}
+				let import_future = self.pool.submit_one(&best_block_id, uxt);
+				let import_future = import_future
+					.then(move |import_result| {
+						match import_result {
+							Ok(_) => report_handle.report_peer(who, reputation_change),
```

Comment:
So this is now going to be async, I hope we do handle non-existing peers correctly.

Reply:
Yes - reporting is done using `mpsc` channel anyway => it has to be safe.